storage/indexer/src/event_v2_translator.rs

```diff
@@ -0,0 +1,1361 @@
+        TokenEventStoreV1Resource, TokenMutation, TokenMutationEvent, TokenOfferEvent,
+        TokenResource, TokenStoreResource, TokenWithdraw, TokenWithdrawEvent, Transfer,
+        TransferEvent, UnlimitedSupplyResource, UriMutation, UriMutationEvent, WithdrawEvent,
+        BURN_EVENT_TYPE, BURN_TOKEN_EVENT_TYPE, BURN_TOKEN_TYPE, BURN_TYPE, CANCEL_OFFER_TYPE,
+        CLAIM_TYPE, COIN_DEPOSIT_TYPE, COIN_REGISTER_EVENT_TYPE, COIN_REGISTER_TYPE,
+        COIN_WITHDRAW_TYPE, COLLECTION_DESCRIPTION_MUTATE_EVENT_TYPE,
+        COLLECTION_DESCRIPTION_MUTATE_TYPE, COLLECTION_MAXIMUM_MUTATE_EVENT_TYPE,
+        COLLECTION_MAXIMUM_MUTATE_TYPE, COLLECTION_MUTATION_EVENT_TYPE, COLLECTION_MUTATION_TYPE,
+        COLLECTION_URI_MUTATE_EVENT_TYPE, COLLECTION_URI_MUTATE_TYPE, CREATE_COLLECTION_EVENT_TYPE,
+        CREATE_COLLECTION_TYPE, CREATE_TOKEN_DATA_EVENT_TYPE, DEFAULT_PROPERTY_MUTATE_EVENT_TYPE,
+        DEFAULT_PROPERTY_MUTATE_TYPE, DEPOSIT_EVENT_TYPE, DESCRIPTION_MUTATE_EVENT_TYPE,
+        DESCRIPTION_MUTATE_TYPE, KEY_ROTATION_EVENT_TYPE, KEY_ROTATION_TYPE,
+        MAXIMUM_MUTATE_EVENT_TYPE, MAXIMUM_MUTATE_TYPE, MINT_EVENT_TYPE, MINT_TOKEN_EVENT_TYPE,
+        MINT_TOKEN_TYPE, MINT_TYPE, MUTATE_PROPERTY_MAP_TYPE, MUTATE_TOKEN_PROPERTY_MAP_EVENT_TYPE,
+        OFFER_TYPE, OPT_IN_TRANSFER_EVENT_TYPE, OPT_IN_TRANSFER_TYPE, ROYALTY_MUTATE_EVENT_TYPE,
+        ROYALTY_MUTATE_TYPE, TOKEN_CANCEL_OFFER_EVENT_TYPE, TOKEN_CLAIM_EVENT_TYPE,
+        TOKEN_DATA_CREATION_TYPE, TOKEN_DEPOSIT_EVENT_TYPE, TOKEN_DEPOSIT_TYPE,
+        TOKEN_MUTATION_EVENT_TYPE, TOKEN_MUTATION_TYPE, TOKEN_OFFER_EVENT_TYPE,
+        TOKEN_WITHDRAW_EVENT_TYPE, TOKEN_WITHDRAW_TYPE, TRANSFER_EVENT_TYPE, TRANSFER_TYPE,
+        URI_MUTATION_EVENT_TYPE, URI_MUTATION_TYPE, WITHDRAW_EVENT_TYPE,
+    },
+    contract_event::{ContractEventV1, ContractEventV2},
+    event::EventKey,
+    state_store::{state_key::StateKey, TStateView},
+    DummyCoinType,
+};
+use bytes::Bytes;
+use dashmap::DashMap;
+use move_core_types::{
+    account_address::AccountAddress,
+    language_storage::{StructTag, TypeTag},
+    move_resource::MoveStructType,
+};
+use once_cell::sync::Lazy;
+use std::{collections::HashMap, str::FromStr, sync::Arc};
+
+pub trait EventV2Translator: Send + Sync {
+    fn translate_event_v2_to_v1(
+        &self,
+        v2: &ContractEventV2,
+        engine: &EventV2TranslationEngine,
+    ) -> Result<ContractEventV1>;
+}
+
+pub struct EventV2TranslationEngine {
+    pub main_db_reader: Arc<dyn DbReader>,
+    pub internal_indexer_db: Arc<DB>,
+    // Map from event type to translator
+    pub translators: HashMap<TypeTag, Box<dyn EventV2Translator + Send + Sync>>,
+    event_sequence_number_cache: DashMap<EventKey, u64>,
+}
+
+impl EventV2TranslationEngine {
+    pub fn new(main_db_reader: Arc<dyn DbReader>, internal_indexer_db: Arc<DB>) -> Self {
+        let translators: HashMap<TypeTag, Box<dyn EventV2Translator + Send + Sync>> = [
```

Comment:
well.. the `.clone` and `Box::new` can be done in a `map()`, no? 

(Guess it's too messy to type the translator constructor, I'm gonna approve)

Reply:
Yeah, it does seem tricky to get the typing to work...